docs/examples/bayesian_optimisation.py

```diff
@@ -0,0 +1,737 @@
+# black-box objective function at this point, and append the new observation to our dataset.
+#
+# Below we plot the posterior distribution of the surrogate model, along with the sample
+# drawn from the model, and the minimiser of this sample returned from the optimiser,
+# which we denote with a star.
+
+
+# %%
+def plot_bayes_opt(
+    posterior: gpx.Module,
+    sample: FunctionalSample,
+    dataset: gpx.Dataset,
+    queried_x: ScalarFloat,
+) -> None:
+    plt_x = jnp.linspace(0, 1, 1000).reshape(-1, 1)
+    forrester_y = forrester(plt_x)
+    sample_y = sample(plt_x)
+
+    latent_dist = posterior.predict(plt_x, train_data=dataset)
+    predictive_dist = posterior.likelihood(latent_dist)
+
+    predictive_mean = predictive_dist.mean()
+    predictive_std = predictive_dist.stddev()
+
+    fig, ax = plt.subplots()
+    ax.fill_between(
+        plt_x.squeeze(),
+        predictive_mean - 2 * predictive_std,
+        predictive_mean + 2 * predictive_std,
+        alpha=0.2,
+        label="Two sigma",
+        color=cols[1],
+    )
+    ax.plot(
+        plt_x,
+        predictive_mean - 2 * predictive_std,
+        linestyle="--",
+        linewidth=1,
+        color=cols[1],
+    )
+    ax.plot(
+        plt_x,
+        predictive_mean + 2 * predictive_std,
+        linestyle="--",
+        linewidth=1,
+        color=cols[1],
+    )
+    ax.plot(
+        plt_x,
+        forrester_y,
+        label="Forrester Function",
+        color=cols[0],
+        linestyle="--",
+        linewidth=2,
+    )
+    ax.plot(plt_x, predictive_mean, label="Predictive Mean", color=cols[1])
+    ax.plot(plt_x, sample_y, label="Posterior Sample")
+    ax.scatter(dataset.X, dataset.y, label="Observations", color=cols[2], zorder=2)
+    ax.scatter(
+        queried_x,
+        sample(queried_x),
+        label="Posterior Sample Optimum",
+        marker="*",
+        color=cols[3],
+        zorder=3,
+    )
+    ax.legend(loc="center left", bbox_to_anchor=(0.975, 0.5))
+    plt.show()
+
+
+plot_bayes_opt(opt_posterior, approx_sample, D, x_star)
+
+# %% [markdown]
+# At this point we can update our model with the newly augmented dataset, and repeat the
+# whole process until some stopping criterion is met. Below we repeat this process for 10
+# iterations, printing out the queried point and the value of the black-box function at
+# each iteration.
+
+# %%
+bo_iters = 10
+
+# Set up initial dataset
+initial_x = tfp.mcmc.sample_halton_sequence(
+    dim=1, num_results=initial_sample_num, seed=key, dtype=jnp.float64
+).reshape(-1, 1)
+initial_y = forrester(initial_x)
+D = gpx.Dataset(X=initial_x, y=initial_y)
+
+for i in range(bo_iters):
+    key, subkey = jr.split(key)
+
+    # Generate optimised posterior using previously observed data
+    mean = gpx.mean_functions.Zero()
+    kernel = gpx.kernels.Matern52()
+    prior = gpx.Prior(mean_function=mean, kernel=kernel)
+    opt_posterior = generate_optimised_posterior(D, prior, subkey)
+
```

Comment:
maybe could just have a whole fit_model method that just takes in D?

Reply:
I had originally planned to do this, but in the second example end up having to use a slightly different prior (with two active dimensions and a slightly higher variance) so felt like being able to pass a prior was a bit easier in the end.